libraries/AP_Scripting/lua_bindings.cpp

```diff
@@ -450,6 +450,77 @@ int AP_HAL__I2CDevice_read_registers(lua_State *L) {
     return success;
 }
 
+int AP_HAL__I2CDevice_write_registers(lua_State *L) {
+    const int args = lua_gettop(L);
+    bool multi_byte;
+    if (args == 2) {
+        multi_byte = false;
+    } else if (args == 3) {
+        multi_byte = true;
+    } else {
+        return luaL_argerror(L, args, "expected 1 or 2 arguments");
+    }
+
+    // stack = [i2c, addr, {table?}]
+    AP_HAL::I2CDevice  *ud = *check_AP_HAL__I2CDevice(L, 1);
+    const uint8_t addr = get_uint8_t(L, 2);
+
+    uint16_t write_size = 1;
+
+    if (multi_byte) {
+        // stack = [i2c, addr, {table}]
+        const size_t table_length_raw = lua_rawlen(L, 3);
+        if (table_length_raw > UINT16_MAX) {
+            return luaL_argerror(L, 3, "table's length out of range");
+        } else {
+            const uint16_t table_length = static_cast<uint16_t>(table_length_raw);
+            write_size += table_length;
+        }
+    }
+
+    uint8_t *buffer = (uint8_t*)luaM_malloc(L, write_size);
+    buffer[0] = addr;
+
+    if (write_size > 1) {
+        lua_pushnil(L);
+        // stack = [i2c, addr, {table}, nil]
+
+        while (lua_next(L, 3) != 0) {
+            // stack = [i2c, addr, {table}, index, value]
+            int isnum_index;
+            const lua_Integer index_raw = lua_tointegerx(L, 4, &isnum_index);
+            if (!isnum_index || (index_raw < 0) || (index_raw > UINT16_MAX)) {
+                luaM_free(L, buffer);
+                return luaL_argerror(L, 3, "table's index out of range");
+            }
+            uint16_t index = static_cast<uint16_t>(index_raw);
+
+            int isnum_value;
+            const lua_Integer value_raw = lua_tointegerx(L, 5, &isnum_value);
+            if (!isnum_value || (value_raw < 0) || (value_raw > UINT8_MAX)) {
+                luaM_free(L, buffer);
+                return luaL_argerror(L, 3, "table's value out of range");
+            }
+            uint8_t value = static_cast<uint8_t>(value_raw);
+
+            buffer[index] = value;
```

Comment:
We need to check were not writing past the end of the buffer, index here could be larger than write size. I would feel happyer with for loop up to write size, less chance of getting stuck in a while forever.

Reply:
The `buffer` is created from 
```
    uint16_t write_size = 1;
...
    const size_t table_length_raw = lua_rawlen(L, 3);
...
    write_size += table_length;
```

So I don't think it should be possible to write over the buffer length. Unless the table mutates between the length check and the loop, or the `rawlen` is broken.
But ok, I'll change the code.

Reply:
Your probably right but its not obviously correct. The for loop is much clearer, also has the advantage that you don't have to pull the index at all.

Reply:
I've updated the code. I left the code to pull the index from the stack just to make sure our index aligns with Lua's.